lib/ai_bot/tools/web_browser.rb

```diff
@@ -0,0 +1,114 @@
+# frozen_string_literal: true
+
+module DiscourseAi
+  module AiBot
+    module Tools
+      class WebBrowser < Tool
+        def self.signature
+          {
+            name: name,
+            description:
+              "Visits a web page, retrieves the HTML content, extracts the main content, converts it to plain text, and returns the result.",
+            parameters: [
+              {
+                name: "url",
+                description: "The URL of the web page to visit.",
+                required: true,
+                type: "string",
+              },
+            ],
+          }
+        end
+
+        def self.name
+          "web_browser"
+        end
+
+        def url
+          return @url if defined?(@url)
+          @url = parameters[:url]
+          @url = "https://#{@url}" if !@url.start_with?("http")
+
+          @url
+        end
+
+        def invoke(_bot_user, llm)
+          send_http_request(url, follow_redirects: true) do |response|
+            if response.code == "200"
+              html = read_response_body(response)
+              text = extract_main_content(html)
+              text = truncate(text, max_length: 50_000, percent_length: 0.3, llm: llm)
+              return { url: response.uri.to_s, text: text.strip }
+            else
+              return { url: url, error: "Failed to retrieve the web page: #{response.code}" }
+            end
+          end
+
+          { url: url, error: "Failed to retrieve the web page" }
+        end
+
+        def description_args
+          { url: url }
+        end
+
+        private
+
+        def extract_main_content(html)
+          doc = Nokogiri.HTML(html)
+
+          # Remove script, style, and comments
+          doc.search("script, style, comment").remove
+
+          # Find the main content container
+          main_content = find_main_content(doc)
+
+          # If main content not found, fallback to the body
+          main_content ||= doc.at("body")
+
+          # Extract text from the main content
+          buffer = +""
+          nodes_to_text(main_content, buffer)
```

Comment:
This is interesting, I would have thought you needed to do this for `buffer` to be updated?

```ruby
buffer = nodes_to_text(main_content, buffer)
```

So the original string is modified inside the method?

Reply:
yes it I like the efficiency of just appending to a buffer, so we need to pass around, otherwise so many string copies ...